### PR TITLE
Fix gradient of expectation value for exact state with complex parameters

### PR DESCRIPTION
**Pull request by yannra**
Hi all,
I noticed that there was a small error in the computation of the gradient of expectation values for the exact state when complex parameters are used in the model. This PR fixes that.

The issue can for example be seen with the following little script comparing the computed gradient with one obtained by numdifftools (yes, I know, it's a little bit strange to use numdifftools to verify the gradients in code based on jax but setting it up with numdifftools was just quicker than making the expectation value evaluation differentiable with jax).

```python
import netket as nk
import numpy as np
import numdifftools as nd
import jax
from jax.nn.initializers import normal

g = nk.graph.Chain(6)
hi = nk.hilbert.Spin(0.5, N=g.n_nodes, total_sz=0.0)
ha = nk.operator.Heisenberg(hi, g, sign_rule=False)
model = nk.models.RBM(dtype=complex, use_visible_bias=False, use_hidden_bias=False, kernel_init=normal(stddev=1.))

vs = nk.vqs.ExactState(hi, model=model, seed=jax.random.PRNGKey(123))
vs_sampled = nk.vqs.MCState(nk.sampler.MetropolisExchange(hi, graph=g), model=model, n_samples=1000000)

""" We are doing a little bit of reshaping and stacking to map the complex
parameters to a flattened array of real values (two times the total number of params). This way,
we can evaluate the gradient with numdifftool and compare."""

kernel_shape = vs.parameters["Dense"]["kernel"].shape
def calculate_expect(parameters):
    kernel = parameters[:len(parameters)//2] + 1.j*parameters[len(parameters)//2:]
    vs.parameters = {"Dense" : {"kernel" : kernel.reshape(kernel_shape)}}
    return vs.expect(ha).mean.real

def calculate_grad_flattened(parameters):
    kernel = parameters[:len(parameters)//2] + 1.j*parameters[len(parameters)//2:]
    vs.parameters = {"Dense" : {"kernel" : kernel.reshape(kernel_shape)}}
    grad_flattened = vs.expect_and_grad(ha)[1]["Dense"]["kernel"].flatten()
    return 2 * np.concatenate((grad_flattened.real, grad_flattened.imag))

""" Just to ensure that the sign definitions are in-line with the rest of NetKet,
we also compute a sampled expectation value."""
def calculate_grad_sampled_flattened(parameters):
    kernel = parameters[:len(parameters)//2] + 1.j*parameters[len(parameters)//2:]
    vs_sampled.parameters = {"Dense" : {"kernel" : kernel.reshape(kernel_shape)}}
    grad_flattened = vs_sampled.expect_and_grad(ha)[1]["Dense"]["kernel"].flatten()
    return 2 * np.concatenate((grad_flattened.real, grad_flattened.imag))

init_pars = vs.parameters["Dense"]["kernel"].flatten()
init_pars_split = np.concatenate((init_pars.real, init_pars.imag))

# gradient as calculated by NetKet
grad_nk = calculate_grad_flattened(init_pars_split)
grad_nk_sampled = calculate_grad_sampled_flattened(init_pars_split)

# Gradient computed by numdifftools
grad_nd = nd.Gradient(calculate_expect)(init_pars_split)

print(max(abs(grad_nd - grad_nk))) # returns large values without the fix in this PR
print(np.mean(abs(grad_nk_sampled - grad_nk)**2)) # returns large values without the fix in this PR
```

The script above just verifies the gradient by comparing the one obtained by calling `vs.expect_and_grad(ha)` with the numerical gradient of `vs.expect(ha)` as obtained with numdifftools.
WIthout the fix of this pull request, it can be seen that the gradients are wrong when complex parameters are used, i.e. there is a significant deviation between the different gradients (which then also results in weird behaviour in the optimization of the exact state with complex parameters).

Incidentally, I realised that the gradient with respect to complex parameters is defined differently in NetKet as compared to jax which might be a little bit confusing for people (happy to open an issue if this is wanted): NetKet gives the complex conjugated version of the gradient which jax would usually return (in the sense that if you do gradient descent with the jax gradients, you need to apply complex conjugation to the complex-valued gradients returned by jax, the NetKet version can directly be used for GD). This discrepancy will probably also lead to issues when using the interface to evaluate gradients for non-hermitian operators (which presumably has never really been tested for models with complex parameters?) as this just evaluates the gradient by jax-differentiating on the full expectation value evaluation. The fix in this PR explicitly assumes Hermitian operators but I guess this is fine as there is no clear definition how a gradient would be defined if the operator is non-hermitian and complex valued models are used (as this can give complex-valued expectation values).

I hope this all makes sense, let me know if you require further info.

Best wishes,
Yannic

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Hey @yannra , thanks a lot for your contribution!

The ExactState functionality is quite new, so there might be a few bugs, indeed. I'd like to ask @nikita-astronaut to check if he agrees with this change (to me it looks correct).

Could you also contribute a test to check that the computed gradient is correct? In 'test/variational/test_exact.py' there are tests already but evidently not enough.

**Comment by femtobit** [comment]
> Incidentally, I realised that the gradient with respect to complex parameters is defined differently in NetKet as compared to jax which might be a little bit confusing for people (happy to open an issue if this is wanted): NetKet gives the complex conjugated version of the gradient which jax would usually return

Yes, it is true and in principle intended that the gradient methods return F = ∂O/∂θ* which is equal to (∂O/∂θ)* for real O and where θ are the variational parameters. This is so that F is the direction of steepest ascent and thus can be used for optimization directly without conjugation. I think this is not actually highlighted by the docs of `expect_and_grad`, in which case they definitely should be improved.

I am not very familiar with the requirements of non-Hermitian physics, but I assume there is no general reason for preferring one of ∂O/∂θ or ∂O/∂θ* (as there is no direct notion of steepest ascent for a strictly complex quantity O). If I am not mistaken, such non-Hermitian operators can still be fully treated in NetKet using a real-parameter model (by treating real and imaginary parts as separate variables, even if the model uses them for complex arithmetic internally), right @yannra?

**Comment by PhilipVinc** [comment]
Hmm, I had quickly read the code without going through the comment, but what @femtobit is saying is right...

Let me get back to you

**Comment by yannra** [comment]
> I am not very familiar with the requirements of non-Hermitian physics, but I assume there is no general reason for preferring one of ∂O/∂θ or ∂O/∂θ* (as there is no direct notion of steepest ascent for a strictly complex quantity O). If I am not mistaken, such non-Hermitian operators can still be fully treated in NetKet using a real-parameter model (by treating real and imaginary parts as separate variables, even if the model uses them for complex arithmetic internally), right @yannra?

yes, that is really what I meant. Just wanted to point out that my fix explicitly assumes Hermitian operators (which is pretty much the same as for the MCState interface). From the comments in the code, it appears that the expectation value calculation for non-hermitian operators is very experimental at the moment and I also wouldn't know how an optimization of the state would work if the expectation value is complex-valued.

Just added the derivative tests for complex valued models (kept it pretty minimal, just shout if you want to see more tests).

**Comment by PhilipVinc** [comment]
I won't shout but I might ask.
I love more tests.

**Comment by gcarleo** [comment]
thanks @yannra for spotting this! just remember to check the style with black (see here https://www.netket.org/docs/contributing.html#python-coding-style) otherwise the automatic tests won't pass 

Concerning the different definition of gradients, yes we are aware of this, actually even Jax is considering adopting the same convention we use (see discussion in google/jax#4891 ), but it's not clear if this will happen any time soon.

**Comment by yannra** [comment]
Oups, many apologies, forgot to run black (should all be fixed now). I also added a couple of more tests using the exact state with different machines which should cover a few more cases.

File: test/variational/test_exact_state.py [hunk 1]
```diff
@@ -122,13 +130,13 @@ def test_qutip_conversion(vstate):
     np.testing.assert_allclose(q_obj.data.todense(), ket.reshape(q_obj.shape))
 
 
-def test_derivatives_agree():
+@pytest.mark.parametrize("machine", machines.values())
```
Comment (PhilipVinc) [inline test/variational/test_exact_state.py:133]: This gives it a more readable name when running under pytest

```suggestion
@pytest.mark.parametrize("machine", [pytest.param(ma, id=name) for name, ma in machines.items()])
```

File: test/variational/test_exact_state.py [hunk 2]
```diff
@@ -54,6 +56,12 @@
     kernel_init=normal(stddev=0.1),
     hidden_bias_init=normal(stddev=0.1),
 )
+machines["model:(C->C)"] = RBM(
+    alpha=10,
+    dtype=complex,
+    kernel_init=normal(stddev=1.0),
+    hidden_bias_init=normal(stddev=1.0),
+)
```
Comment (PhilipVinc) [inline test/variational/test_exact_state.py:64]: Why are you adding another RBM with much higher stddev and alpha? 
Isn't the test by the one above enough?

Comment (PhilipVinc) [reply test/variational/test_exact_state.py:64]: Besides, I think this stddev is too large for `alpha=10`. This might be why tests are failing

File: test/variational/test_exact_state.py [hunk 3]
```diff
@@ -181,15 +189,21 @@ def central_diff_grad(func, x, eps, *args, dtype=None):
...
 
 @common.skipif_mpi
-@pytest.mark.parametrize("L,n_iterations,h", [(4, 100, 1), (6, 100, 2), (8, 100, 3)])
+@pytest.mark.parametrize(
+    "L,n_iterations,h",
+    [(4, 100, 1), (6, 100, 2), (8, 100, 3)],
+)
+@pytest.mark.parametrize(
+    "machine",
+    machines.values(),
```
Comment (PhilipVinc) [inline test/variational/test_exact_state.py:198]: here too
```suggestion
    "machine", [pytest.param(ma, id=name) for name, ma in machines.items()]
```

File: test/variational/test_exact_state.py [hunk 4]
```diff
@@ -122,13 +130,13 @@ def test_qutip_conversion(vstate):
     np.testing.assert_allclose(q_obj.data.todense(), ket.reshape(q_obj.shape))
 
 
-def test_derivatives_agree():
+@pytest.mark.parametrize("machine", machines.values())
+def test_derivatives_agree(machine):
     err = 1e-3
     g = nk.graph.Hypercube(length=10, n_dim=1, pbc=True)
```
Comment (PhilipVinc) [inline test/variational/test_exact_state.py:136]: I know it wasn't you, but let's reduce slightly the size so the tests run faster, and use Chain which is equivalent but more readable
```suggestion
    g = nk.graph.Chain(length=8, pbc=True)
```

**Review by PhilipVinc — COMMENTED** [review]
>Interesting that jax also wants to change the convention for the complex valued gradients, so I guess then there really is no point in touching that (tbh I also prefer the convention as is used in NetKet at the moment). It was just something I stumbled across when checking the derivatives...

You can mention it in the docstring if you want. We probably should anyway...

@yannra it seems like that 2 tests are not passing... 1 is probably just some numerical threshold to be increased, while the other one test_TFIM_energy_strictly_decreases[machine2-8-100-3] ) looks like another numerical threshold that went wrong.

I think they can both be fixed by reducing the initial variance to `0.1` for the new machine you just added.

File: test/variational/test_exact_state.py (hunk 2 above)
Comment (PhilipVinc) [reply test/variational/test_exact_state.py]: I run the tests 20 times on my machine and it failed 3 times with `normal(stddev=1.0)` but never with `normal(stddev=0.1)`. I really think this should be decreased.

File: test/variational/test_exact_state.py [hunk 5]
```diff
@@ -54,6 +56,12 @@
     kernel_init=normal(stddev=0.1),
     hidden_bias_init=normal(stddev=0.1),
 )
+machines["model:(C->C) (alt)"] = RBM(
```
Comment (PhilipVinc) [inline test/variational/test_exact_state.py:59]:
```suggestion
machines["model:(C->C,α=5)"] = RBM(
```

File: test/variational/test_exact_state.py (hunk 2 above)
Comment (yannra) [reply test/variational/test_exact_state.py]: The initial reason why I added another (more complex) machine was that sometimes the derivatives can become relatively small for an initial state which only fluctuates a tiny amount around a uniform state. This way, the gradient checks might go through by accident even if the derivatives are wrong. Just wanted to add a model which is significantly further away from a uniform state just to be sure that the derivatives are right. I do appreciate that the standard deviation was a little bit too large for things still to be numerically stable, will decrease that a bit which should fix things.
(Tests always passed on my machine so I didn't realize these issues before looking at the github test outputs...)

**Review by PhilipVinc — APPROVED** [review]
thanks again @yannra for your contribution!

**Comment by PhilipVinc** [comment]
Last failure is unrelated (it's the flaky qgt test)
